File: packages/core/src/ids.ts
import { randomBytes } from "node:crypto";

/** Crockford-style base32, lowercase, without ambiguous characters (i, l, o, u). */
const ALPHABET = "0123456789abcdefghjkmnpqrstvwxyz";

export type IdPrefix =
  | "tnt" // tenant
  | "usr" // user (the human principal)
  | "agt" // agent identity
  | "ups" // upstream tool backend
  | "grt" // delegation grant
  | "pol" // policy
  | "apr" // approval request
  | "call" // tool call
  | "evt"; // audit event

/** 20 chars over a 32-symbol alphabet ≈ 100 bits of entropy. */
export function newId(prefix: IdPrefix, size = 20): string {
  const bytes = randomBytes(size);
  let out = "";
  for (const b of bytes) out += ALPHABET[b % 32];
  return `${prefix}_${out}`;
}

File: src/toolgate/core/ids.py
import secrets
from typing import Literal

# Crockford-style base32, lowercase, without ambiguous characters (i, l, o, u).
_ALPHABET = "0123456789abcdefghjkmnpqrstvwxyz"

IdPrefix = Literal["tnt", "usr", "agt", "ups", "grt", "pol", "apr", "call", "evt"]


def new_id(prefix: IdPrefix, size: int = 20) -> str:
    """20 chars over a 32-symbol alphabet ~= 100 bits of entropy."""
    body = "".join(_ALPHABET[b % 32] for b in secrets.token_bytes(size))
    return f"{prefix}_{body}"
